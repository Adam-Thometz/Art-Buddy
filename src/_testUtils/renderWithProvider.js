import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';

// import renderWithRouter from './renderWithRouter';
import rootReducer from '../_redux/reducers/rootReducer';

function withProvider({ children, store }) {
  return (
    <Provider store={store}>
      <MemoryRouter>
        {children}
      </MemoryRouter>
    </Provider>
  )
};

function renderWithProvider(component, options) {
  const store = (options && options.store)
    ? options.store
    : rootReducer;
  return render(component, { wrapper: (args) =>
    withProvider({
      ...args,
      store
    }), ...options
  });
};

export default renderWithProvider;