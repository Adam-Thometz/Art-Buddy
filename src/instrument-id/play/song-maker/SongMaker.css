.SongMaker-button-instrument-wrapper {
  display: flex;
  width: 85%;
  margin: 0 auto;
  height: 60vh;
  align-items: center;
}

.SongMaker-buttons {
  width: 33%;
}

.SongMaker-buttons button {
  width: 50%;
}