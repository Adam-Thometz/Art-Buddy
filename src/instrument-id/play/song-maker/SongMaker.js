import React from 'react';

import { useSelector, useDispatch } from 'react-redux';
import { addInstrument, selectInstrument } from '../../../_redux/actions/insturmentIdActions';

import './SongMaker.css';

import Button from '../../../_components/button/Button';
import Dropdown from '../../../_components/dropdown/Dropdown';
import Icon from '../../../_components/icon/Icon';

import { addInstrumentIcon } from '../../_icons/iconImports';
import { instrumentOptions, melodyOptions } from './dropdownOptions';
import Popup from '../../../_components/popup/Popup';
import SaveSong from './save-song/SaveSong';

const SongMaker = () => {
  const instruments = useSelector(state => state.instrumentId.instruments);
  const dispatch = useDispatch()
  
  const handleAddInstrument = e => {
    const id = e.currentTarget.id;
    dispatch(addInstrument(id));
  }
  
  const handleSelectInstrument = e => {
    const id = e.currentTarget.id;
    const instrument = e.target.id
    dispatch(selectInstrument({ id, instrument }));
  }

  const savePopup = (
    <Popup
      title='SAVE SONG'
      trigger={<Button colorId={3}>SAVE</Button>}
      popup={<SaveSong />}
    />
  )

  const instrumentDisplay = instruments.map((instrument, i) => (
    !instrument ? (
      <Icon
        icon={addInstrumentIcon}
        text='ADD INSTRUMENT'
        size='69px'
        id={i}
        onClick={handleAddInstrument}
      />
    ) : (
      <div className='SongMaker-options'>
        <Dropdown
          id={i}
          labelText='INSTRUMENT'
          options={instrumentOptions}
          onClick={handleSelectInstrument}
        />
        {instrument.instrument ? (
          <Dropdown
            id={i}
            labelText='MELODY'
            options={melodyOptions}
          />
        ) : null}
      </div>
    )
  ));

  return (
    <div className='SongMaker'>
      <p>HULLO IM SONG MAKER MAKE A SONG PLZ</p>
      <hr/>
      <div className='SongMaker-button-instrument-wrapper'>
        <div className='SongMaker-buttons'>
          <Button colorId={0}>PLAY</Button>
          {savePopup}
        </div>
        <div className='SongMaker-instruments'>
          {instrumentDisplay}
        </div>
      </div>
    </div>
  );
};

export default SongMaker;