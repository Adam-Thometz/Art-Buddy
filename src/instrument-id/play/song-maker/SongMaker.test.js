import React from "react";
import renderWithProvider from "../../../_testUtils/renderWithProvider";
import { screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import SongMaker from "./SongMaker";

describe('SongMaker component', () => {
  it('renders without crashing', () => {
    renderWithProvider(<SongMaker />);
  });

  it('matches snapshot', () => {
    const { asFragment } = renderWithProvider(<SongMaker />);
    expect(asFragment()).toMatchSnapshot();
  });

  it('selects melodies', () => {
    renderWithProvider(<SongMaker />);
    const instrumentDropdown = screen.getByText('INSTRUMENT');
    userEvent.click(instrumentDropdown);
    const instrumentOption = screen.getByText('VIOLIN');
    userEvent.click(instrumentOption);
    expect(screen.getByText('MELODY')).toBeInTheDocument();
    const melodyDropdown = screen.getByText('MELODY');
    userEvent.click(melodyDropdown);
    const melodyOption = screen.getByText('BABY SHARK');
    userEvent.click(melodyOption);
    expect(screen.queryByText('MELODY')).not.toBeInTheDocument();
  });
  
  it('selects rhythms', () => {
    renderWithProvider(<SongMaker />);
    const addInstrumentBtns = screen.getAllByText('ADD INSTRUMENT')
    userEvent.click(addInstrumentBtns[0])
    expect(screen.getAllByText('INSTRUMENT')[1]).toBeInTheDocument();

    const instrumentDropdown = screen.getAllByText('INSTRUMENT')[1];
    userEvent.click(instrumentDropdown);
    const instrumentOption = screen.getByText('DRUM SET');
    userEvent.click(instrumentOption);
    expect(screen.getByText('RHYTHM')).toBeInTheDocument();

    const rhythmDropdown = screen.getByText('RHYTHM');
    userEvent.click(rhythmDropdown);
    const rhythmOption = screen.getByText('REGULAR');
    userEvent.click(rhythmOption);
    expect(screen.queryByText('RHYTHM')).not.toBeInTheDocument();
  });
  
  it('saves a song', async () => {
    renderWithProvider(<SongMaker />);
    const saveBtn = screen.getAllByText('SAVE')[0];
    userEvent.click(saveBtn);
    expect(screen.getByText('SAVE SONG')).toBeInTheDocument();

    const input = screen.getByLabelText('Song Title');
    userEvent.type(input, 'Regular Baby');
    const submit = screen.getAllByText('SAVE')[1];
    userEvent.click(submit);
    expect(screen.getByText('Song Saved!')).toBeInTheDocument();

    const close = screen.getByText('CLOSE');
    userEvent.click(close);
    const savedSongs = screen.getByText('SAVED SONGS');
    userEvent.click(savedSongs);
    expect(screen.getByText('Regular Baby')).toBeInTheDocument();
  });
});