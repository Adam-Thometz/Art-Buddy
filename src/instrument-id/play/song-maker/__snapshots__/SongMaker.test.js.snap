// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SongMaker component matches snapshot 1`] = `
<DocumentFragment>
  <header
    class="WindowNavbar-games"
  >
    <h1
      class="WindowNavbar-label"
    >
      Song Maker
    </h1>
    <div
      class="WindowNavbar-corner"
    >
      <div
        class="SavedSongsIcon"
      >
        <figure
          class="Icon"
          data-testid=""
        >
          <img
            alt=""
            src="file"
            style="height: 49px; width: 49px;"
          />
          <figcaption
            class="Icon-text"
            style="font-size: 1rem;"
          >
            SAVED SONGS
          </figcaption>
        </figure>
      </div>
    </div>
  </header>
  <section
    class="SongMaker-button-instrument-wrapper"
  >
    <div
      class="PlaySave"
    >
      <button
        class="Button regular"
        style="border-color: #89eb0a;"
      >
        PLAY
      </button>
      <div>
        <button
          class="Button regular"
          style="border-color: #4d4f5e;"
        >
          SAVE
        </button>
      </div>
    </div>
    <div
      class="InstrumentDisplay"
    >
      <div
        class="InstrumentDisplay-options"
      >
        <div
          class="Dropdown"
        >
          <span
            class="Dropdown-label"
            style="z-index: 5;"
          >
            INSTRUMENT 
            <img
              alt=""
              src="file"
            />
          </span>
        </div>
        <span
          class="InstrumentDisplay-remove"
          id="0"
        >
          remove
        </span>
      </div>
      <figure
        class="Icon"
        data-testid=""
        id="1"
      >
        <img
          alt=""
          src="file"
          style="height: 69px; width: 69px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1.75rem;"
        >
          ADD INSTRUMENT
        </figcaption>
      </figure>
      <figure
        class="Icon"
        data-testid=""
        id="2"
      >
        <img
          alt=""
          src="file"
          style="height: 69px; width: 69px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1.75rem;"
        >
          ADD INSTRUMENT
        </figcaption>
      </figure>
      <figure
        class="Icon"
        data-testid=""
        id="3"
      >
        <img
          alt=""
          src="file"
          style="height: 69px; width: 69px; cursor: pointer;"
        />
        <figcaption
          class="Icon-text"
          style="font-size: 1.75rem;"
        >
          ADD INSTRUMENT
        </figcaption>
      </figure>
    </div>
  </section>
</DocumentFragment>
`;
